src/QMCHamiltonians/EwaldTools.h

```diff
@@ -0,0 +1,382 @@
+{
+  real_t v = 0.0;
+  IntVec iv;
+  if(zero)
+  {
+    // Directly sum over the full space
+    for(int_t i = -nmax[0]; i < nmax[0] + 1; ++i)
+      for(int_t j = -nmax[1]; j < nmax[1] + 1; ++j)
+        for(int_t k = -nmax[2]; k < nmax[2] + 1; ++k)
+        {
+          iv[0] = i;
+          iv[1] = j;
+          iv[2] = k;
+          v += function(iv);
+        }
+  }
+  else
+  {
+    // Sum half spaces around zero in z
+    for(int_t k = 1; k < nmax[2] + 1; ++k)
+      for(int_t i = -nmax[0]; i < nmax[0] + 1; ++i)
+        for(int_t j = -nmax[1]; j < nmax[1] + 1; ++j)
+        {
+          iv[0] = i;
+          iv[1] = j;
+          iv[2] = k;
+          v += function(iv);
+          iv[2] = -k;
+          v += function(iv);
+        }
+    // Sum half planes around zero in y
+    iv[2] = 0;
+    for(int_t j = 1; j < nmax[1] + 1; ++j)
+      for(int_t i = -nmax[0]; i < nmax[0] + 1; ++i)
+        {
+          iv[0] = i;
+          iv[1] = j;
+          v += function(iv);
+          iv[1] = -j;
+          v += function(iv);
+        }
+    // Sum half lines around zero in x
+    iv[1] = 0;
+    for(int_t i = 1; i < nmax[0] + 1; ++i)
+    {
+      iv[0] = i;
+      v += function(iv);
+      iv[0] = -i;
+      v += function(iv);
+    }
+  }
+  return v;
+}
+
+
+
+class AnisotropicEwald
+{
+private:
+
+  RealMat A;
+  RealMat B;
+  ChargeArray Q;
+  real_t tol;
+
+  size_t N;
+  real_t volume;
+  real_t QQmax;
+
+  real_t kappa_madelung;
+  real_t rconst_madelung;
+  real_t kconst_madelung;
+  real_t kfactor_madelung;
+
+  real_t kappa_ewald;
+  real_t rconst_ewald;
+  real_t kconst_ewald;
+  real_t kfactor_ewald;
+
+  real_t madelung_constant;
+  real_t madelung_energy;
+
+  IntVec nmax_anisotropic;
+
+public:
+
+  AnisotropicEwald(const RealMat& A_in, const ChargeArray& Q_in, real_t tol_in = 1e-10,real_t kappa_in=-1.0)
+    : A(A_in), Q(Q_in), tol(tol_in), nmax_anisotropic(0)
+  {
+
+    // Reciprocal lattice vectors
+    B = 2 * M_PI * transpose(inverse(A));
+
+    // Set Ewald sum constants
+    volume = std::abs(det(A));
+    // Madelung constants
+    if(kappa_in<0.0)
+      kappa_madelung = getKappaMadelung(volume);
+    else
+      kappa_madelung = kappa_in;
+    rconst_madelung  = kappa_madelung;
+    kconst_madelung  = -1. / (4 * std::pow(kappa_madelung, 2));
+    kfactor_madelung = 4 * M_PI / volume;
+    // Ewald constants
+    if(kappa_in<0.0)
+      kappa_ewald = getKappaEwald(volume);
+    else
+      kappa_ewald = kappa_in;
+    rconst_ewald  = 1. / (std::sqrt(2.) * kappa_ewald);
+    kconst_ewald  = -std::pow(kappa_ewald, 2) / 2;
+    kfactor_ewald = 4 * M_PI / volume;
+
+    // Number of particles
+    N = Q.size();
+
+    // Find maximum self-interaction charge product
+    QQmax = 0.0;
+    for (size_t i = 0; i < N; ++i)
+      QQmax = std::max(std::abs(Q[i] * Q[i]), QQmax);
+
+    // Calculate and store Madelung energy
+    madelung_energy = madelungEnergy();
+  }
+
+
+  real_t madelungConstant()
```

Comment:
The naming already follows the coding conventions, as in the manual:

```
27.3.5 (Member) function names
Function names should start with a lowercase character and have a capital letter for each new word.
```

Adding `calculate` to every function that calculates something is IMO trite and lexically wasteful.

Will add const.

Reply:
`madelungConstant()` is not clear whether it to access a stored value in the class or run a calculation. When I name functions, evaluate/calculateXXX() means it can be costly. getXXX() means cheap. Probably in your case, it doesn't matter much.